code/go/internal/validator/semantic/validate_kibana_tag_duplicates.go

```diff
@@ -0,0 +1,126 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package semantic
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/fs"
+	"path"
+
+	"github.com/elastic/package-spec/v3/code/go/internal/fspath"
+	"github.com/elastic/package-spec/v3/code/go/pkg/specerrors"
+	"gopkg.in/yaml.v3"
+)
+
+const (
+	tagType = "tag"
+)
+
+type packageSpecTag struct {
+	Attributes struct {
+		Name string `yaml:"name"`
+	} `yaml:"attributes"`
+	Type string `yaml:"type"`
+}
+
+type sharedTag struct {
+	Name string `yaml:"text"`
+}
+
+// ValidateKibanaTagDuplicates checks for duplicate Kibana tag names
+// between the kibana/tags.yml file and the tags defined in the package's kibana/tag/*.json files.
+// It returns a list of validation errors if any duplicates are found.
+func ValidateKibanaTagDuplicates(fsys fspath.FS) specerrors.ValidationErrors {
+	tagMap, errs := getKibanaTagsYMLMap(fsys)
+	if len(errs) > 0 {
+		return errs
+	}
+
+	errs = validateKibanaJSONTags(fsys, tagMap)
+	return errs
+}
+
+func getKibanaTagsYMLMap(fsys fspath.FS) (map[string]struct{}, specerrors.ValidationErrors) {
+	tagMap := make(map[string]struct{})
+	tagsPath := path.Join("kibana", "tags.yml")
+	// Collect all tags defined in the kibana/tags.yml file.
+	b, err := fs.ReadFile(fsys, tagsPath)
+	if err != nil {
+		if errors.Is(err, fs.ErrNotExist) {
+			return tagMap, nil
+		}
+		return nil, specerrors.ValidationErrors{specerrors.NewStructuredError(err, specerrors.UnassignedCode)}
+	}
+	var sharedKibanaTags []sharedTag
+	err = yaml.Unmarshal(b, &sharedKibanaTags)
+	if err != nil {
+		return nil, specerrors.ValidationErrors{specerrors.NewStructuredError(err, specerrors.UnassignedCode)}
+	}
+
+	errs := make(specerrors.ValidationErrors, 0)
+	// Check for duplicate tag names in the kibana/tags.yml file.
+	for _, tag := range sharedKibanaTags {
+		if _, exists := tagMap[tag.Name]; exists {
+			errs = append(errs, specerrors.NewStructuredError(
+				fmt.Errorf("duplicate tag name '%s' found in %s", tag.Name, tagsPath), specerrors.UnassignedCode))
+			continue
+		}
+		tagMap[tag.Name] = struct{}{}
+	}
+	return tagMap, errs
+}
+
+func validateKibanaJSONTags(fsys fspath.FS, tagMap map[string]struct{}) specerrors.ValidationErrors {
+	// Collect all tags used in the package.
+	tagDir := path.Join("kibana", "tag")
```

Comment:
Can tags be embedded directly in dashboards? 🤔 Probably don't, but it would be worth checking.

Reply:
There are elements that can be embedded in dashboards, but they are mostly the panels and the searches. Probably tags can only be included by reference. Thanks for checking :+1: